src/platform/web/ui/general/LazyListView.js

```diff
@@ -0,0 +1,251 @@
+            this.renderCount + topGrow + bottomGrow,
+            this.bottomCount - bottomGrow,
+        );
+    }
+
+    totalSize() {
+        return this.topCount + this.renderCount + this.bottomCount;
+    }
+
+    normalize(idx) {
+        /*
+        map index from list to index in rendered range
+        eg: if the index range of this._list is [0, 200] and we have rendered
+        elements in range [50, 60] then index 50 in list must map to index 0
+        in DOM tree/childInstance array.
+        */
+        return idx - this.topCount;
+    }
+}
+
+export class LazyListView extends ListView {
+    constructor({itemHeight, overflowMargin = 5, overflowItems = 20,...options}, childCreator) {
+        super(options, childCreator);
+        this._itemHeight = itemHeight;
+        this._overflowMargin = overflowMargin;
+        this._overflowItems = overflowItems;
+    }
+
+    _getVisibleRange() {
+        const length = this._list ? this._list.length : 0;
+        const scrollTop = this._parent.scrollTop;
+        const topCount = Math.min(Math.max(0, Math.floor(scrollTop / this._itemHeight)), length);
+        const itemsAfterTop = length - topCount;
+        const visibleItems = this._height !== 0 ? Math.ceil(this._height / this._itemHeight) : 0;
+        const renderCount = Math.min(visibleItems, itemsAfterTop);
+        const bottomCount = itemsAfterTop - renderCount;
+        return new ItemRange(topCount, renderCount, bottomCount);
+    }
+
+    _renderIfNeeded(forceRender = false) {
+        /*
+        forceRender only because we don't optimize onAdd/onRemove yet.
+        Ideally, onAdd/onRemove should only render whatever has changed + update padding + update renderRange
+        */
+        const range = this._getVisibleRange();
+        const intersectRange = range.expand(this._overflowMargin);
+        const renderRange = range.expand(this._overflowItems);
+        // only update render Range if the new range + overflowMargin isn't contained by the old anymore 
+        // or if we are force rendering
+        if (forceRender || !this._renderRange.contains(intersectRange)) {
+            this._renderRange = renderRange;
+            this._renderElementsInRange();
+        }
+    }
+
+    async _initialRender() {
+        /*
+        Wait two frames for the return from mount() to be inserted into DOM.
+        This should be enough, but if this gives us trouble we can always use
+        MutationObserver.
+        */
+        await new Promise(r => requestAnimationFrame(r));
+        await new Promise(r => requestAnimationFrame(r));
+
+        this._height = this._parent.clientHeight;
+        if(this._height === 0) { throw "LazyListView could not calculate parent height."}
+        const range = this._getVisibleRange();
+        const renderRange = range.expand(this._overflowItems);
+        this._renderRange = renderRange;
+        this._renderElementsInRange();
+    }
+
+    _renderItems(items) {
```

Comment:
since this method being called from anywhere but _renderElementsInRange would break things (as _renderElementsInRange clears the DOM), should we merge the two?

Reply:
Have merged the two for the time being. The follow up PR will change the render code anyway.